regexec.c

```diff
@@ -366,25 +370,33 @@ S_regcppop(pTHX_ regexp *rex, U32 *maxopenparen_p _pDEPTH)
                 PTR2UV(rex->offs)
             );
     );
-    paren = *maxopenparen_p;
-    for ( ; i > 0; i -= REGCP_PAREN_ELEMS) {
-        SSize_t tmps;
-        rex->offs[paren].start_tmp = SSPOPINT;
-        rex->offs[paren].start = SSPOPIV;
-        tmps = SSPOPIV;
-        if (paren <= rex->lastparen)
-            rex->offs[paren].end = tmps;
-        DEBUG_BUFFERS_r( Perl_re_exec_indentf( aTHX_
-            "    \\%" UVuf ": %" IVdf "(%" IVdf ")..%" IVdf "%s\n",
-            depth,
-            (UV)paren,
-            (IV)rex->offs[paren].start,
-            (IV)rex->offs[paren].start_tmp,
-            (IV)rex->offs[paren].end,
-            (paren > rex->lastparen ? "(skipped)" : ""));
-        );
-        paren--;
-    }
+    /* substract remaining elements from the stack */
+    PL_savestack_ix -= i;
+
+    /* static assert that offs struc size is not less than stack elem size */
+    STATIC_ASSERT_STMT(sizeof(*rex->offs) >= sizeof(*PL_savestack));
+
+    /* calculate number of offs/capture groups stored */
+    i = (i * sizeof(*PL_savestack)) / sizeof(*rex->offs);
```

Comment:
I keep staring at this and wondering what happens when `i=1` and `sizeof(PL_savestack) == 4` and `sizeof(*rex->offs) == 12`.  Doesnt it mean that `i` ends ups as `0`? `(1*4)/ 12 == 0`. Is that intended?

I am not necessarily questioning the correctness of this logic, but I do feel it could be a touch more self explaining or use a bit more detail in the comments. It seems that line wants to change the i units of i, but it doesnt explain how the case of a fractional unit is dealt with. Is it just the case that i cannot be 1 when sizeof(rex->ofs) is >= sizeof(PL_savestack)? or is it the case that we always round up and somewhere after we add 1? Its not clear to me if that is what is happening with `*maxopenparen_p - i + 1` in the following statement.

Reply:
It's +1 because it's like that on other places.

Reply:
The comments are helpful, but its still not clear to me from the code alone why `i` ending up as 0 would be ok, or why it would never happen.

Reply:
> I keep staring at this and wondering what happens when `i=1` and `sizeof(PL_savestack) == 4` and `sizeof(*rex->offs) == 12`. Doesnt it mean that `i` ends ups as `0`? `(1*4)/ 12 == 0`. Is that intended?
> 
> I am not necessarily questioning the correctness of this logic, but I do feel it could be a touch more self explaining or use a bit more detail in the comments. It seems that line wants to change the i units of i, but it doesnt explain how the case of a fractional unit is dealt with. Is it just the case that i cannot be 1 when sizeof(rex->ofs) is >= sizeof(PL_savestack)? or is it the case that we always round up and somewhere after we add 1? Its not clear to me if that is what is happening with `*maxopenparen_p - i + 1` in the following statement.

This scenario is impossible since earlier (in regcppush) we align the elements.